### DISCUSSION TITLE
How to include newlines in CDATA block?

### DISCUSSION THREAD

**Discussion by arklumpus** [AvaloniaUI/Avalonia#12635] [Q&A] [answered]
Hi, I'm still updating stuff to Avalonia 11... I suppose there have been some changes to the XAML interpreter that make it swallow newlines in CDATA blocks.

Is this intentional? If so, is there a way to tell the parser to preserve newlines within `CDATA` blocks (and/or everywhere)?

The reason why I would like this to be possible is that I have a control that can be used to render Markdown content in Avalonia. The Markdown source code is set through the `DocumentSource` property of this control; in Avalonia 0.10 it used to be possible to do it like this:

```XAML
<mdc:MarkdownCanvasControl>
      <mdc:MarkdownCanvasControl.DocumentSource>
<![CDATA[
# Markdown title

Markdown text
]]>
      </mdc:MarkdownCanvasControl.DocumentSource>
    </mdc:MarkdownCanvasControl>
```

This allows users to just write the Markdown source within the XAML and have it appear in the previewer.

However, this does not work anymore, because the newlines are swallowed by the XAML parser (even though they are in a `CDATA` block).

**Comment by stevemonaco** [comment, marked as answer]
There have been some whitespace-related commits in XamlX which may have changed behavior. Try setting `<mdc:MarkDownCanvasControl xml:space="preserve">` and see what happens. It works for me when using `TextBlock`:

```xml
<TextBlock xml:space="preserve">
    <TextBlock.Text>
        Testing
        Some more
    </TextBlock.Text>
</TextBlock>
```

Edit: Also works with CDATA.
```xml
<TextBlock xml:space="preserve">
    <TextBlock.Text>
<![CDATA[
# Markdown title

Markdown text
]]>
    </TextBlock.Text>
</TextBlock>
```

**Reply by arklumpus** [reply]
Thanks, that's exactly what I was looking for! I had thought about trying it, but when I started entering `xml:` in the editor, it got autocompleted to `xmlns:`, so I assumed it was not supported... But yes, it does work!

Hoever, `xml:space="preserve"` seems to cause formatting whitespace to be interpreted as the `Content` of the control. For example:

```XAML
<mdc:MarkdownCanvasControl xml:space="preserve">
    <mdc:MarkdownCanvasControl.DocumentSource>
<![CDATA[
# Markdown title

Markdown text
]]>
    </mdc:MarkdownCanvasControl.DocumentSource>
</mdc:MarkdownCanvasControl>
```

Produces a seemingly empty control, because the whole contents of the `MarkDownCanvasControl` (which is a `UserControl` containing a `ScrollViewer` and other stuff) are replaced with a text block containing whitespace.

Instead, this works as expected:

```XAML
<mdc:MarkdownCanvasControl xml:space="preserve"><mdc:MarkdownCanvasControl.DocumentSource>
<![CDATA[
# Markdown title

Markdown text
]]>
    </mdc:MarkdownCanvasControl.DocumentSource></mdc:MarkdownCanvasControl>
```

I easily solved this for my control by making it ignore changes to the `Content` property after it has been initialised, but I'm leaving this here just in case someone else stumbles on this solution.

**Reply by timunie** [reply]
tbh, I'd have a seperate md file in assets or whereever and load it from code behind. Much cleaner imo

**Reply by arklumpus** [reply]
Yeah, definitely... It's nice to have a self-contained example, though

**Reply by timunie** [reply]
https://docs.avaloniaui.net/docs/next/basics/user-interface/assets

**Reply by arklumpus** [reply]
Sorry, I meant self-contained as in "everything is in the same file" 😅

**Reply by jp2masa** [reply]
Did you try applying `xml:space="preserve"` to `<mdc:MarkdownCanvasControl.DocumentSource>`?

**Reply by arklumpus** [reply]
Yes, `System.Xaml.XamlException: 'Attributes aren't allowed on element properties Line 13, position 4.' Line number '13' and line position '4'.`

**Reply by jp2masa** [reply]
One more suggestion, although a bit ugly: did you try wrapping the text in `<x:String xml:space="preserve">`?

**Reply by arklumpus** [reply]
Yeah, `x:String` works!
